topics/reference_guide/text_selection.md

```diff
@@ -0,0 +1,55 @@
+
+## Extending/Shrinking Text Selection
+
+EP: `com.intellij.extendWordSelectionHandler`
+
+Implementing the [`ExtendWordSelectionHandler`](upsource:///platform/lang-api/src/com/intellij/codeInsight/editorActions/ExtendWordSelectionHandler.java) interface and registering it as `com.intellij.extendWordSelectionHandler` EP in your `plugin.xml` allows you to provide additional text ranges to be used when extending or shrinking a selection.
+You then return `true` from the `canSelect(PsiElement)` for the PSI elements that you want to provide additional text-ranges for.
+The IntelliJ Platform will call `select(PsiElement, CharSequence, int, Editor)` for these elements where you can compute additional text ranges and return them as `List<TextRange>`.
+
+See also:
+
+- [Surround With](surround_with.md)
+- [List of minor features](additional_minor_features.md)
+
+### Overview
+
+The two actions [Extend Selection and Shrink Selection](https://www.jetbrains.com/help/idea/working-with-source-code.html#editor_code_selection) in IntelliJ Platform IDEs let you adjust selected text based on the structure of the source code.
+This makes it easy to select not only expressions, blocks, and function definitions, but also code like whole lines or tags in JavaDoc comments.
+
+If you’re implementing a custom language, the IntelliJ Platform will provide you with basic implementations of this EP, allowing you to select code based on your PSI structure and to select whole lines.
+In many cases this is sufficient to provide a good user experience.
+However, sometimes it’s advantageous to provide additional regions that the user may wish to be able to select when extending or shrinking a selection. 
+
+This EP has two methods that need to be implemented:
+
+1. The `canSelect(PsiElement)` method is called on each PSI element, starting from the element at the cursor and walking up each of its parents.
+   By returning `true` for a particular element, you can indicate that further text-ranges should be included for the PSI element.
+2. In the `select(PsiElement, CharSequence, int, Editor)` method, the specific text-ranges within the PSI element of interest are calculated and returned.
+
+### Example Use-Case
+
+A possible use-case for custom language developers is a function call `f(a, b)` where the function call node has its two arguments as children.
+If the cursor was, for example, at argument `a`, extending the selection would first select argument `a` itself and in the next step grow to cover the whole function call.
+However, you might want to select the list of all arguments as an intermediate step.
+This can be achieved by implementing this EP in the following way:
+
+1. Create a class that implements the `ExtendWordSelectionHandler` interface and register it as a `com.intellij.extendWordSelectionHandler` EP in your `plugin.xml`.
```

Comment:
somewhat duplicating intro

Reply:
Yes, but I wanted to be very clear at this point. That's why it says "create a class" and "implement the interface".